### PR TITLE
image_view for ros2

### PR DESCRIPTION
**Pull request by ahuizxc**
already ported image_view for ros2

### CODE REVIEW

**Comment by ahuizxc** [comment]
this error caused by wrong version of ros when building this pr.

File: image_view/CMakeLists.txt [hunk 1]
```diff
@@ -1,90 +1,148 @@
...
+find_package(ament_cmake REQUIRED)
+find_package(rclcpp REQUIRED)
+find_package(rosidl_default_generators REQUIRED)
+find_package(builtin_interfaces REQUIRED)
+find_package(rmw_implementation REQUIRED)
+find_package(cv_bridge REQUIRED)
+find_package(tf2_eigen REQUIRED)
+find_package(image_geometry REQUIRED)
+find_package(image_transport REQUIRED)
+find_package(sensor_msgs REQUIRED)
```
Comment (yechun1) [inline image_view/CMakeLists.txt:24]: sensor_msgs could be removed here

File: image_view/CMakeLists.txt [hunk 2]
```diff
@@ -1,90 +1,148 @@
...
+find_package(stereo_msgs REQUIRED)
+find_package(tf2 REQUIRED)
+find_package(tf2_ros REQUIRED)
+find_package(OpenCV REQUIRED)
+find_package(Boost REQUIRED COMPONENTS signals system filesystem)
```
Comment (yechun1) [inline image_view/CMakeLists.txt]: please try to remove boost, std is recommended in ROS2

File: image_view/CMakeLists.txt [hunk 3]
```diff
@@ -1,90 +1,148 @@
...
+            src/nodelets/disparity_node.cpp
+            src/nodelets/window_thread.cpp)
+
+ament_target_dependencies(${PROJECT_NAME}
+  image_transport
+  image_geometry
+  cv_bridge
+  std_srvs
+  rclcpp
+  sensor_msgs
```
Comment (yechun1) [inline image_view/CMakeLists.txt:48]: standard msgs should be include in rclcpp, needn't add here again.

File: image_view/CMakeLists.txt [hunk 4]
```diff
@@ -1,90 +1,148 @@
...
+  stereo_msgs
+  rmw_implementation
+  cv_bridge
+  OpenCV
+  Boost
```
Comment (yechun1) [inline image_view/CMakeLists.txt:53]: Please try to remove all Boost dependencies

File: image_view/package.xml [hunk 5]
```diff
@@ -1,46 +1,37 @@
-<package>
+<package format="2">
   <name>image_view</name>
   <version>1.12.23</version>
   <description>
-  A simple viewer for ROS image topics. Includes a specialized viewer
-  for stereo + disparity images.
+
```
Comment (yechun1) [inline image_view/package.xml]: seems copied wrong description, please keep original one

File: image_view/package.xml [hunk 6]
```diff
@@ -1,46 +1,37 @@
...
   <author>Patrick Mihelich</author>
   <maintainer email="vincent.rabaud@gmail.com">Vincent Rabaud</maintainer>
   <license>BSD</license>
-  <url>http://www.ros.org/wiki/image_view</url>
-
-  <export>
-    <rosdoc config="rosdoc.yaml" />
-    <nodelet plugin="${prefix}/nodelet_plugins.xml" />
-  </export>
+  <url>http://ros.org/wiki/depth_image_proc</url>
```
Comment (yechun1) [inline image_view/package.xml]: seems copied wrong description, please keep original one

File: image_view/package.xml [hunk 7]
```diff
@@ -1,46 +1,37 @@
...
 
-  <test_depend>rostest</test_depend>
+  <test_depend>ament_cmake_gtest</test_depend>
+  <test_depend>ament_lint_auto</test_depend>
+  <test_depend>ament_lint_common</test_depend>
 
-  <build_depend>camera_calibration_parsers</build_depend>
-  <build_depend version_gte="1.11.13">cv_bridge</build_depend>
-  <build_depend>dynamic_reconfigure</build_depend>
-  <build_depend>gtk2</build_depend>
```
Comment (yechun1) [inline image_view/package.xml:25]: why gtk2 removed here?

File: image_view/src/nodelets/disparity_node.cpp [hunk 8]
```diff
@@ -31,115 +31,81 @@
...
+#include <stereo_msgs/msg/disparity_image.hpp>
 #include <opencv2/highgui/highgui.hpp>
 #include "window_thread.h"
+#include "visibility.h"
+#include <memory>
+#include <string>
 
 #ifdef HAVE_GTK
 #include <gtk/gtk.h>
-
```
Comment (yechun1) [inline image_view/src/nodelets/disparity_node.cpp:43]: any reason to remove these gtk code?

File: image_view/src/nodes/extract_images.cpp [hunk 9]
```diff
@@ -34,84 +34,90 @@
...
-
-    image_transport::ImageTransport it(nh);
-    sub_ = it.subscribe(topic, 1, &ExtractImages::image_cb, this, transport);
+    double sec_per_frame;
+    node->get_parameter_or("sec_per_frame", sec_per_frame, 0.1);
+    sec_per_frame_ = rclcpp::Time(sec_per_frame);
+    std::shared_ptr<image_transport::ImageTransport> it;
+    it.reset(new image_transport::ImageTransport(node));
+    image_transport::TransportHints hints(node, transport);
+    sub_ = it->subscribe(topic, 1, std::bind(&ExtractImages::image_cb, this, _1), node, &hints);
```
Comment (yechun1) [inline image_view/src/nodes/extract_images.cpp]: please use new pointer api and ref to https://github.com/ros-perception/image_pipeline/blob/843b932985acc41889071b1749a1e13d80ea58c2/depth_image_proc/src/nodelets/crop_foremost.cpp#L93

File: image_view/src/nodes/extract_images.cpp [hunk 10]
```diff
@@ -122,33 +128,25 @@ class ExtractImages
...
     }
   }
 };
 
 int main(int argc, char **argv)
 {
-  ros::init(argc, argv, "extract_images", ros::init_options::AnonymousName);
-  ros::NodeHandle n;
-  if (n.resolveName("image") == "/image") {
-    ROS_WARN("extract_images: image has not been remapped! Typical command-line usage:\n"
```
Comment (yechun1) [inline image_view/src/nodes/extract_images.cpp:145]: there is command line usage, please keep the information when port.

File: image_view/src/nodes/image_saver.cpp [hunk 11]
```diff
@@ -133,22 +138,26 @@ class Callbacks {
       return;
 
     // save the CameraInfo
-    if (info) {
-      filename = filename.replace(filename.rfind("."), filename.length(), ".ini");
-      camera_calibration_parsers::writeCalibration(filename, "camera", *info);
-    }
+    // if (info) {
```
Comment (yechun1) [inline image_view/src/nodes/image_saver.cpp:142]: suggest added TODO for later updates

File: image_view/src/nodes/image_saver.cpp [hunk 12]
```diff
@@ -182,45 +189,69 @@ class Callbacks {
...
+  node = rclcpp::Node::make_shared("image_saver");
+
+  std::shared_ptr<image_transport::ImageTransport> it;
+  it.reset(new image_transport::ImageTransport(node));
+  image_transport::TransportHints hints(node);
   Callbacks callbacks;
   // Useful when CameraInfo is being published
-  image_transport::CameraSubscriber sub_image_and_camera = it.subscribeCamera(topic, 1,
+
+  image_transport::CameraSubscriber sub_image_and_camera = it->subscribeCamera(topic, 1,
```
Comment (yechun1) [inline image_view/src/nodes/image_saver.cpp]: please use image_transport::create_camera_subscription instead

**Review by yechun1 — CHANGES_REQUESTED** [review]
Add some tiny comments, would you please help to check and fix? thanks.
Please make as minimal changes as possible for code porting, suggest to add TODO for all code to modify later.

File: image_view/src/nodes/image_view.cpp [hunk 13]
```diff
@@ -120,13 +119,14 @@ static void mouseCb(int event, int x, int y, int flags, void* param)
...
-
-  std::string filename = (g_filename_format % g_count).str();
+  sprintf(buf, format_string.c_str(), g_count);
+  std::string filename = buf;
   if (cv::imwrite(filename, image)) {
     RCLCPP_INFO(node->get_logger(), "Saved image %s", filename.c_str());
     g_count++;
   } else {
-    boost::filesystem::path full_path = boost::filesystem::complete(filename);
+    std::experimental::filesystem::path full_path =
```
Comment (mjcarroll) [inline image_view/src/nodes/image_view.cpp]: This won't work on MacOS, I suggest looking at camera_info_manager/filesystem_helper.hpp

Comment (ahuizxc) [reply image_view/src/nodes/image_view.cpp]: I commented out this feature, and filesystem will add to macOS in the future, when it add to macOS, this function will enable ASAP :).

**Pull request closed without merging**